File: router/services/routing_service.py
"""
Routing Service
Manages routing table and message forwarding between devices
"""

import threading
import time
from typing import Dict, Optional, Set
from enum import Enum
from dataclasses import dataclass

from core.config import RoutingConfig
from core.logger import get_logger
from protocol import PacketCrypto


class InterfaceType(Enum):
    """Types of network interfaces"""
    WIFI = "wifi"
    BLE = "ble"
    ZIGBEE = "zigbee"  # Future expansion
    LORA = "lora"      # Future expansion


@dataclass
class RouteEntry:
    """Routing table entry for a device"""
    node_addr: int
    device_id: str
    interface_type: InterfaceType
    connection_handle: any  # Socket for WiFi, BLE handle for BLE, etc.
    session_crypto: PacketCrypto
    last_seen: float
    connection_info: Dict[str, any]


class RoutingService:
    """
    Service for managing device routing and forwarding
    """
    
    # Address space constants
    ROUTER_ADDRESS = 0x0000
    BROADCAST_ADDRESS = 0xFFFF
    MIN_CLIENT_ADDRESS = 0x0001
    MAX_CLIENT_ADDRESS = 0xFFFE
    
    def __init__(self, config: RoutingConfig):
        self.config = config
        self.logger = get_logger(__name__)
        
        # Routing table
        self.routes: Dict[int, RouteEntry] = {}
        self.device_id_to_addr: Dict[str, int] = {}
        self.next_address = self.MIN_CLIENT_ADDRESS
        
        # Thread safety
        self.lock = threading.RLock()
        
        # Service state
        self.running = False
        
        self.logger.info("Routing service initialized", max_devices=config.max_devices)
    
    def start(self):
        """Start the routing service"""
        with self.lock:
            if self.running:
                return
            
            self.running = True
            self.logger.info("Routing service started")
    
    def stop(self):
        """Stop the routing service"""
        with self.lock:
            if not self.running:
                return
            
            self.running = False
            
            # Clear all routes
            self.routes.clear()
            self.device_id_to_addr.clear()
            
            self.logger.info("Routing service stopped")
    
    def allocate_address(self, device_id: str) -> int:
        """
        Allocate a unique address for a device
        Returns existing address if device already has one
        """
        with self.lock:
            # Check if device already has an address
            if device_id in self.device_id_to_addr:
                existing_addr = self.device_id_to_addr[device_id]
                self.logger.debug(f"Device {device_id} already has address 0x{existing_addr:04X}")
                return existing_addr
            
            # Check device limit
            if len(self.routes) >= self.config.max_devices:
                raise RuntimeError(f"Maximum device limit reached: {self.config.max_devices}")
            
            # Find next available address
            original_next = self.next_address
            while self.next_address in self.routes:
                self.next_address += 1
                if self.next_address > self.MAX_CLIENT_ADDRESS:
                    self.next_address = self.MIN_CLIENT_ADDRESS
                
                # Check if we've wrapped around completely
                if self.next_address == original_next:
                    raise RuntimeError("No available addresses in range")
            
            addr = self.next_address
            self.device_id_to_addr[device_id] = addr
            self.next_address += 1
            
            self.logger.debug(f"Allocated address 0x{addr:04X} for device {device_id}")
            return addr
    
    def add_route(self, device_id: str, interface_type: InterfaceType,
                  connection_handle: any, session_crypto: PacketCrypto,
                  connection_info: Dict[str, any] = None) -> int:
        """
        Add a new route entry
        Returns the assigned address
        """
        with self.lock:
            if not self.running:
                raise RuntimeError("Routing service is not running")
            
            addr = self.allocate_address(device_id)
            
            route = RouteEntry(
                node_addr=addr,
                device_id=device_id,
                interface_type=interface_type,
                connection_handle=connection_handle,
                session_crypto=session_crypto,
                last_seen=time.time(),
                connection_info=connection_info or {}
            )
            
            self.routes[addr] = route
            
            self.logger.info(f"Added route for device {device_id}",
                           address=f"0x{addr:04X}",
                           interface=interface_type.value)
            
            return addr
    
    def remove_route(self, addr: int) -> bool:
        """Remove a route entry by address"""
        with self.lock:
            if addr not in self.routes:
                return False
            
            route = self.routes[addr]
            device_id = route.device_id
            
            del self.routes[addr]
            
            if device_id in self.device_id_to_addr:
                del self.device_id_to_addr[device_id]
            
            self.logger.info(f"Removed route for device {device_id}",
                           address=f"0x{addr:04X}")
            
            return True
    
    def remove_route_by_device_id(self, device_id: str) -> bool:
        """Remove a route entry by device ID"""
        with self.lock:
            if device_id not in self.device_id_to_addr:
                return False
            
            addr = self.device_id_to_addr[device_id]
            return self.remove_route(addr)
    
    def get_route(self, addr: int) -> Optional[RouteEntry]:
        """Get route entry by address"""
        with self.lock:
            return self.routes.get(addr)
    
    def get_route_by_device_id(self, device_id: str) -> Optional[RouteEntry]:
        """Get route entry by device ID"""
        with self.lock:
            if device_id not in self.device_id_to_addr:
                return None
            
            addr = self.device_id_to_addr[device_id]
            return self.routes.get(addr)
    
    def update_last_seen(self, addr: int):
        """Update last seen timestamp for a route"""
        with self.lock:
            if addr in self.routes:
                self.routes[addr].last_seen = time.time()
    
    def get_all_routes(self) -> Dict[int, RouteEntry]:
        """Get all route entries (copy)"""
        with self.lock:
            return self.routes.copy()
    
    def get_routes_by_interface(self, interface_type: InterfaceType) -> Dict[int, RouteEntry]:
        """Get all routes for a specific interface type"""
        with self.lock:
            return {
                addr: route for addr, route in self.routes.items()
                if route.interface_type == interface_type
            }
    
    def cleanup_stale_routes(self, timeout_seconds: int) -> int:
        """
        Remove routes that haven't been seen for timeout_seconds
        Returns number of routes removed
        """
        with self.lock:
            current_time = time.time()
            stale_addresses = []
            
            for addr, route in self.routes.items():
                if current_time - route.last_seen > timeout_seconds:
                    stale_addresses.append(addr)
            
            removed_count = 0
            for addr in stale_addresses:
                if self.remove_route(addr):
                    removed_count += 1
            
            if removed_count > 0:
                self.logger.info(f"Cleaned up {removed_count} stale routes")
            
            return removed_count
    
    def get_device_count(self) -> int:
        """Get total number of connected devices"""
        with self.lock:
            return len(self.routes)
    
    def get_device_count_by_interface(self, interface_type: InterfaceType) -> int:
        """Get number of devices connected via specific interface"""
        with self.lock:
            return sum(1 for route in self.routes.values()
                      if route.interface_type == interface_type)
    
    def is_address_available(self, addr: int) -> bool:
        """Check if an address is available for allocation"""
        with self.lock:
            return (self.MIN_CLIENT_ADDRESS <= addr <= self.MAX_CLIENT_ADDRESS and
                    addr not in self.routes)
    
    def get_address_for_device(self, device_id: str) -> Optional[int]:
        """Get the address assigned to a device, if any"""
        with self.lock:
            return self.device_id_to_addr.get(device_id)

File: router/src/services/routing_service.cpp
#include "services/routing_service.hpp"
#include "services/packet_monitor_service.hpp"
#include "core/config.hpp"
#include "core/logger.hpp"
#include "protocol/protocol.hpp"
#include <algorithm>
#include <set>

namespace mita
{
    namespace services
    {

        RoutingService::RoutingService(const core::RoutingConfig &config)
            : config_(config), logger_(core::get_logger("RoutingService")), next_address_(MIN_CLIENT_ADDRESS), packet_monitor_(nullptr)
        {

            logger_->info("Routing service initialized",
                          core::LogContext().add("max_devices", config_.max_devices).add("device_timeout", config_.device_timeout).add("cleanup_interval", config_.cleanup_interval));
        }

        RoutingService::~RoutingService()
        {
            stop();
        }

        void RoutingService::start()
        {
            if (running_.exchange(true))
            {
                return; // Already running
            }

            logger_->info("Routing service started");
        }

        void RoutingService::stop()
        {
            if (!running_.exchange(false))
            {
                return; // Already stopped
            }

            // Clear all routes
            {
                std::unique_lock<std::shared_mutex> routes_lock(routes_mutex_);
                std::lock_guard<std::shared_mutex> device_map_lock(device_map_mutex_);
                std::lock_guard<std::mutex> allocation_lock(allocation_mutex_);

                routing_table_.clear();
                device_to_address_.clear();
                allocated_addresses_.clear();
                next_address_ = MIN_CLIENT_ADDRESS;
            }

            logger_->info("Routing service stopped");
        }

        uint16_t RoutingService::add_device(const std::string &device_id, core::TransportType interface_type,
                                            void *connection_handle)
        {
            // Check if device already exists
            {
                std::shared_lock<std::shared_mutex> lock(device_map_mutex_);
                auto it = device_to_address_.find(device_id);
                if (it != device_to_address_.end())
                {
                    logger_->warning("Device already exists in routing table",
                                     core::LogContext().add("device_id", device_id).add("existing_address", it->second));
                    return it->second;
                }
            }

            // Assign new address
            uint16_t address = assign_address(device_id);
            if (address == 0)
            {
                logger_->error("Failed to assign address to device",
                               core::LogContext().add("device_id", device_id));
                return 0;
            }

            // Create route entry
            RouteEntry entry(address, device_id, interface_type, connection_handle);

            // Add to routing table
            add_route_entry(address, entry);

            logger_->info("Device added to routing table",
                          core::LogContext().add("device_id", device_id).add("address", address).add("interface_type", static_cast<int>(interface_type)));

            return address;
        }

        bool RoutingService::remove_device(const std::string &device_id)
        {
            uint16_t address = 0;

            // Find device address
            {
                std::shared_lock<std::shared_mutex> lock(device_map_mutex_);
                auto it = device_to_address_.find(device_id);
                if (it == device_to_address_.end())
                {
                    logger_->warning("Device not found in routing table",
                                     core::LogContext().add("device_id", device_id));
                    return false;
                }
                address = it->second;
            }

            return remove_device_by_address(address);
        }

        bool RoutingService::remove_device_by_address(uint16_t address)
        {
            std::string device_id;

            // Remove from routing table
            {
                std::unique_lock<std::shared_mutex> lock(routes_mutex_);
                auto it = routing_table_.find(address);
                if (it == routing_table_.end())
                {
                    return false;
                }
                device_id = it->second.device_id;
                routing_table_.erase(it);
            }

            // Remove from device mapping
            {
                std::unique_lock<std::shared_mutex> lock(device_map_mutex_);
                device_to_address_.erase(device_id);
            }

            // Release address
            release_address(address);

            logger_->info("Device removed from routing table",
                          core::LogContext().add("device_id", device_id).add("address", address));

            return true;
        }

        const RouteEntry *RoutingService::get_route(uint16_t address) const
        {
            std::shared_lock<std::shared_mutex> lock(routes_mutex_);
            auto it = routing_table_.find(address);
            return (it != routing_table_.end()) ? &it->second : nullptr;
        }

        const RouteEntry *RoutingService::get_route_by_device_id(const std::string &device_id) const
        {
            uint16_t address = get_device_address(device_id);
            if (address == 0)
            {
                return nullptr;
            }
            return get_route(address);
        }

        std::map<uint16_t, RouteEntry> RoutingService::get_all_routes() const
        {
            std::shared_lock<std::shared_mutex> lock(routes_mutex_);
            return routing_table_;
        }

        uint16_t RoutingService::assign_address(const std::string &device_id)
        {
            if (!config_.auto_assign_addresses)
            {
                logger_->error("Automatic address assignment is disabled");
                return 0;
            }

            std::lock_guard<std::mutex> lock(allocation_mutex_);

            // Check device limit
            if (allocated_addresses_.size() >= static_cast<size_t>(config_.max_devices))
            {
                logger_->error("Maximum device limit reached",
                               core::LogContext().add("max_devices", config_.max_devices));
                return 0;
            }

            uint16_t address = find_free_address();
            if (address == 0)
            {
                logger_->error("No free addresses available");
                return 0;
            }

            allocated_addresses_.insert(address);

            // Update device mapping
            {
                std::unique_lock<std::shared_mutex> lock(device_map_mutex_);
                device_to_address_[device_id] = address;
            }

            return address;
        }

        bool RoutingService::is_address_available(uint16_t address) const
        {
            if (address < MIN_CLIENT_ADDRESS || address > MAX_CLIENT_ADDRESS)
            {
                return false;
            }

            std::lock_guard<std::mutex> lock(allocation_mutex_);
            return allocated_addresses_.find(address) == allocated_addresses_.end();
        }

        uint16_t RoutingService::get_device_address(const std::string &device_id) const
        {
            std::shared_lock<std::shared_mutex> lock(device_map_mutex_);
            auto it = device_to_address_.find(device_id);
            return (it != device_to_address_.end()) ? it->second : 0;
        }

        void RoutingService::set_session_crypto(const std::string &device_id,
                                                std::shared_ptr<protocol::PacketCrypto> crypto)
        {
            const RouteEntry *route = get_route_by_device_id(device_id);
            if (!route)
            {
                logger_->warning("Cannot set session crypto for unknown device",
                                 core::LogContext().add("device_id", device_id));
                return;
            }

            std::unique_lock<std::shared_mutex> lock(routes_mutex_);
            auto it = routing_table_.find(route->node_addr);
            if (it != routing_table_.end())
            {
                it->second.session_crypto = crypto;
            }
        }

        std::shared_ptr<protocol::PacketCrypto> RoutingService::get_session_crypto(const std::string &device_id) const
        {
            const RouteEntry *route = get_route_by_device_id(device_id);
            return route ? route->session_crypto : nullptr;
        }

        bool RoutingService::route_packet(const protocol::ProtocolPacket &packet)
        {
            uint16_t dest_address = packet.get_dest_addr();

            // Capture packet for monitoring
            if (packet_monitor_)
            {
                // Determine direction based on source/destination
                std::string direction = "forwarded";
                if (packet.get_source_addr() == ROUTER_ADDRESS)
                {
                    direction = "outbound";
                }
                else if (dest_address == ROUTER_ADDRESS)
                {
                    direction = "inbound";
                }

                // Get transport type from route (default to WiFi if unknown)
                core::TransportType transport = core::TransportType::WIFI;
                const RouteEntry *route = get_route(packet.get_source_addr());
                if (route)
                {
                    transport = route->interface_type;
                }

                packet_monitor_->capture_packet(packet, direction, transport);
            }

            if (dest_address == BROADCAST_ADDRESS)
            {
                return broadcast_packet(packet) > 0;
            }

            return forward_to_device(dest_address, packet);
        }

        bool RoutingService::forward_to_device(uint16_t dest_address, const protocol::ProtocolPacket &packet)
        {
            const RouteEntry *route = get_route(dest_address);
            if (!route)
            {
                logger_->warning("No route found for destination",
                                 core::LogContext().add("dest_address", dest_address));
                return false;
            }

            // Note: Actual packet forwarding is handled by the transport layer/device management
            // This service only validates routes and updates timestamps
            logger_->debug("Route found for packet forwarding",
                           core::LogContext().add("device_id", route->device_id).add("dest_address", dest_address).add("packet_size", packet.get_payload().size()));

            // Update last seen
            {
                std::unique_lock<std::shared_mutex> lock(routes_mutex_);
                auto it = routing_table_.find(dest_address);
                if (it != routing_table_.end())
                {
                    it->second.last_seen = std::chrono::steady_clock::now();
                }
            }

            return true;
        }

        int RoutingService::broadcast_packet(const protocol::ProtocolPacket &packet)
        {
            auto routes = get_all_routes();
            int sent_count = 0;

            for (const auto &[address, route] : routes)
            {
                if (forward_to_device(address, packet))
                {
                    sent_count++;
                }
            }

            logger_->debug("Broadcast packet sent",
                           core::LogContext().add("recipients", sent_count).add("packet_size", packet.get_payload().size()));

            return sent_count;
        }

        int RoutingService::cleanup_stale_routes(std::chrono::seconds timeout_seconds)
        {
            auto cutoff_time = std::chrono::steady_clock::now() - timeout_seconds;
            std::vector<uint16_t> stale_addresses;

            // Find stale routes
            {
                std::shared_lock<std::shared_mutex> lock(routes_mutex_);
                for (const auto &[address, route] : routing_table_)
                {
                    if (route.last_seen < cutoff_time)
                    {
                        stale_addresses.push_back(address);
                    }
                }
            }

            // Remove stale routes
            int removed_count = 0;
            for (uint16_t address : stale_addresses)
            {
                if (remove_device_by_address(address))
                {
                    removed_count++;
                }
            }

            if (removed_count > 0)
            {
                logger_->info("Cleaned up stale routes",
                              core::LogContext().add("removed_count", removed_count));
            }

            return removed_count;
        }

        void RoutingService::update_device_last_seen(const std::string &device_id)
        {
            const RouteEntry *route = get_route_by_device_id(device_id);
            if (!route)
            {
                return;
            }

            std::unique_lock<std::shared_mutex> lock(routes_mutex_);
            auto it = routing_table_.find(route->node_addr);
            if (it != routing_table_.end())
            {
                it->second.last_seen = std::chrono::steady_clock::now();
            }
        }

        size_t RoutingService::get_device_count() const
        {
            std::shared_lock<std::shared_mutex> lock(routes_mutex_);
            return routing_table_.size();
        }

        std::vector<std::string> RoutingService::get_connected_device_ids() const
        {
            std::vector<std::string> device_ids;

            std::shared_lock<std::shared_mutex> lock(routes_mutex_);
            device_ids.reserve(routing_table_.size());

            for (const auto &[address, route] : routing_table_)
            {
                device_ids.push_back(route.device_id);
            }

            return device_ids;
        }

        // Private methods

        uint16_t RoutingService::find_free_address()
        {
            // Start from next_address_ and wrap around
            uint16_t start_address = next_address_;

            do
            {
                if (allocated_addresses_.find(next_address_) == allocated_addresses_.end())
                {
                    uint16_t result = next_address_;

                    // Advance for next allocation
                    next_address_++;
                    if (next_address_ > MAX_CLIENT_ADDRESS)
                    {
                        next_address_ = MIN_CLIENT_ADDRESS;
                    }

                    return result;
                }

                next_address_++;
                if (next_address_ > MAX_CLIENT_ADDRESS)
                {
                    next_address_ = MIN_CLIENT_ADDRESS;
                }

            } while (next_address_ != start_address);

            return 0; // No free address found
        }

        void RoutingService::release_address(uint16_t address)
        {
            std::lock_guard<std::mutex> lock(allocation_mutex_);
            allocated_addresses_.erase(address);
        }

        void RoutingService::add_route_entry(uint16_t address, const RouteEntry &entry)
        {
            std::unique_lock<std::shared_mutex> lock(routes_mutex_);
            routing_table_[address] = entry;
        }

        void RoutingService::remove_route_entry(uint16_t address)
        {
            std::unique_lock<std::shared_mutex> lock(routes_mutex_);
            routing_table_.erase(address);
        }

    } // namespace services
} // namespace mita